File: src/api/scaffold/template.ts
import { invoke } from "@tauri-apps/api";

export async function listTemplate(kind: String): Promise<Version[]> {
    let result: ListTemplateReply = await invoke("list_templates", { kind });
    return result.data;
}

export async function readTemplate(version: Version, kind: string): Promise<string> {
    let result: ReadTemplateReply = await invoke("read_template", { version, kind });
    return result.data;
}

export async function saveTemplate(version: Version, kind: string, template: string) {
    await invoke("save_template", { version, kind, template });
}

export async function fetchOfficalTemplate(kind: string): Promise<{ dev: string, qc: string }> {
    kind = mapKind(kind);
    const versions = await listTemplate(kind);
    const officalDev = versions.filter(v => v.offical && v.role === "dev");
    const officalQc = versions.filter(v => v.offical && v.role === "qc");
    const devVersion = officalDev[officalDev.length - 1].name;
    const qcVersion = officalQc[officalQc.length - 1].name;
    const dev = await readTemplate({ name: devVersion, offical: true, role: "dev" }, kind);
    const qc = await readTemplate({ name: qcVersion, offical: true, role: "qc" }, kind);
    return { dev, qc };
}

function mapKind(kind: string): string {
    switch (kind) {
        case "ADaM":
            return "ADAM";
        case "TFLs":
            return "TFL";
        default:
            return kind;
    }
}


export interface ListTemplateReply {
    data: Version[]
}

export interface ReadTemplateReply {
    data: string
}

export interface Version {
    name: string,
    role: string,
    offical: boolean,
}

File: src-tauri/src/scaffold/template.rs
use std::{env, path::Path};

use scaffold::{Kind, Version, VersionManager, VersionManagerParam};
use serde::Serialize;

use crate::config::{PRIVATE_TEMPLATE, TEMPLATE};

#[tauri::command]
pub fn list_templates(kind: Kind) -> Result<ListTemplateReply, String> {
    let manager = init_manager(&kind);
    match manager.list_templates() {
        Ok(versions) => Ok(ListTemplateReply { data: versions }),
        Err(err) => Err(err.to_string()),
    }
}

#[tauri::command]
pub fn read_template(version: Version, kind: Kind) -> Result<ReadTemplateReply, String> {
    let manager = init_manager(&kind);
    match manager.read_template(&version) {
        Ok(template) => Ok(ReadTemplateReply {
            data: String::from_utf8(template).unwrap(),
        }),
        Err(err) => Err(err.to_string()),
    }
}

#[tauri::command]
pub fn save_template(version: Version, kind: Kind, template: String) -> Result<(), String> {
    let manager = init_manager(&kind);
    match manager.save_template(&version, template.as_bytes()) {
        Ok(_) => Ok(()),
        Err(err) => Err(err.to_string()),
    }
}

fn init_manager(kind: &Kind) -> VersionManager {
    let offical =
        env::var(TEMPLATE).expect(&format!("Error: missing environment variable {}", TEMPLATE));
    let private = env::var(PRIVATE_TEMPLATE).expect(&format!(
        "Error: missing environment variable {}",
        PRIVATE_TEMPLATE
    ));
    let offical_path = Path::new(&offical);
    let private_path = Path::new(&private);
    VersionManager::new(&VersionManagerParam {
        offical_path,
        private_path,
        kind: kind.to_owned(),
    })
}

#[derive(Debug, Serialize)]
pub struct ListTemplateReply {
    data: Vec<Version>,
}

#[derive(Debug, Serialize)]
pub struct ReadTemplateReply {
    data: String,
}
